### PR TITLE
[1/n] Merged fine-tuning dataset: grammar + samsum

### PR DESCRIPTION
**Pull request by RdoubleA**
#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

As discussed in the RFC in #1186, we will merged instruct and chat datasets to the following unified pipeline that can better support multimodal:
1. `message_transform` to create `List[Message]` from dataset with full flexibility on columns, ad-hoc modifications, etc. For multimodal, additionally images are loaded from the path
2. `prompt_template` as a optional way to add structured text around specific roles in the list of messages
3. `model_transform` that takes the list of messages and tokenizes it. For multimodal, it will additionally apply model-specific image transforms to the images associated with the sample

**For ease of review, we will stage this as multiple moderate-sized PRs**. This PR creates the unified dataset class, and refactors grammar and samsum to start off with. As a result, a few key changes were made:
- New `FinetuneDataset` (not married to the name, best I can think of) class with the unified pipeline with associated unit test
- Add `ToInputOutputMessages` (open to better names) which provides a generic message transform that takes input column -> user message, output column -> assistant message
- Refactor grammar and samsum datasets. They now create a message transform using `ToInputOutputMessages`, add a default prompt template that can be changed, and use the new `FinetuneDataset` class
- Move `Message` from `_types.py` to `_messages.py`. This file will now contain everything related to `Message`, including generic message transforms (like the ones in `_converters.py` which will eventually migrate here)
- Make all tokenizers also double as `Transform`s for use in the `model_transform` argument
- New unified `PromptTemplate` interface that merges functionality of `instructTemplate` and `ChatFormat`. Refactored grammar and summarize templates to use a common `CustomPromptTemplate` class that takes in any template and formats a list of messages accordingly. It also covers the ability to specify a custom template from configs, which we were missing. More will be discussed in an upcoming tutorial update.

#### Test plan
- [x] update live docs and check rendering
- [x] unit tests for `FinetuneDataset`, grammar, samsum
- [x] unit tests for `ToInputOutputMessages`
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [x] compare loss curves of grammar, samsum to original versions on main
image
image

### CODE REVIEW

**Comment by RdoubleA** [comment]
> Where are my deprecation warnings? :)

will be added in an upcoming PR... there's a lot stacked on this one :)

**Review by felipemello1 — COMMENTED** [review]
first pass, not thinking too hard about the design. The main thing that caught my attention was trying to understand what exactly FinetuneDataset is. The first line makes it sound like its a Base or a General purpose dataset that other datasets would use, but Chat and Instruct dont, so it is not immediately clear to me when to use one or the other. I hope it makes sense.

File: torchtune/data/_prompt_templates.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+        Returns:
+            The formatted list of messages
+        """
+        pass
+
+
+class CustomPromptTemplate(PromptTemplate):
+    """
+    Define a quick custom prompt template by passing in a dictionary mapping role to
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply torchtune/data/_prompt_templates.py:42]: haha yeah, quickly define is what I meant

File: torchtune/data/_prompt_templates.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+        System: {content}\\n
+        User: {content}\\n
+        Assistant: {content}\\n
+        Tool: {content}\\n
+
+    You can define the template as follows::
+
+        template = {
+            "system": ("System: ", "\\n"),
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply torchtune/data/_prompt_templates.py:54]: fair point. will update

File: torchtune/data/_prompt_templates.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            "user": ("User: ", "\\n"),
+            "assistant": ("Assistant: ", "\\n"),
+            "ipython": ("Tool: ", "\\n"),
+        }
+
+    Once instantiated, you must call the prompt template on a list of messages. It
+    will return the same list of messages updated with the template.
+
+    Note:
+        Any tags prepended/appended to the assistant message will be included
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply torchtune/data/_prompt_templates.py:64]: Generally, only assistant messages are used in loss calculation. unless train_on_input is True, in which case both user and assistant messages are used. system and ipython are never included in loss (as far as I'm aware)

> What I am understanding is that assistant will be included

This is indeed the correct understanding. I can make this a bit more clear.

File: torchtune/datasets/_finetune.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+from datasets import load_dataset
+from torch.utils.data import Dataset
+from torchtune.data import CROSS_ENTROPY_IGNORE_IDX, PromptTemplate
+from torchtune.modules.transforms import Transform
+
+
+class FinetuneDataset(Dataset):
+    """
+    Dataset class for creating instruct, chat, tool, or multimodal datasets for fine-tuning.
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply torchtune/datasets/_finetune.py:19]: No. This will replace both Instruct and ChatDataset. We are essentially merging the two, while also adding support for multimodal.

I haven't thought of a better name than FinetuneDataset. Maybe `TuneDataset` or `TokenizedDataset`?

File: torchtune/data/_prompt_templates.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+            messages (List[Message]): a single conversation, structured as a list
+                of :class:`~torchtune.data.Message` objects
+
+        Returns:
+            The formatted list of messages
+        """
+        pass
+
+
+class CustomPromptTemplate(PromptTemplate):
```
Comment (joecummings) [inline torchtune/data/_prompt_templates.py:40]: I'm in favor of renaming this to `PromptTemplate` and call the other thing `PromptTemplateInterface`.

The "Custom" part of prompt template feels redundant. Like what's the difference between a CustomPromptTemplate and a regular PromptTemplate? I see here by looking at the code that it's b/c one is an interface, but that's not clear from the names. This should be evident before having to go to the docs.

File: torchtune/datasets/_finetune.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+import numpy as np
+
+from datasets import load_dataset
+from torch.utils.data import Dataset
+from torchtune.data import CROSS_ENTROPY_IGNORE_IDX, PromptTemplate
+from torchtune.modules.transforms import Transform
+
+
+class FinetuneDataset(Dataset):
```
Comment (joecummings) [inline torchtune/datasets/_finetune.py:17]: Is this actually only for SFT? If so, what makes it SFT specific? 

And if it is, then should we rename to `SFTDataset`????

File: torchtune/data/_prompt_templates.py (hunk 5 above)
Comment (joecummings) [reply torchtune/data/_prompt_templates.py:40]: We also use the Interface naming for protocols with our recipes so it makes sense.

Comment (RdoubleA) [reply torchtune/data/_prompt_templates.py:40]: that makes sense to me, although if a user needs a bit more custom behavior that's not offered by `CustomPromptTemplate`, I wouldn't want them to accidentally inherit from `PromptTemplate` instead of `PromptTemplateInterface`.

File: torchtune/datasets/_finetune.py (hunk 6 above)
Comment (RdoubleA) [reply torchtune/datasets/_finetune.py:17]: It is SFT specific in the sense that it isn't Preference or TextCompletion. i.e., it covers the previous instruct, chat, and eventual multimodal datasets but it does not cover 1) chosen/rejected messaged in Preference, 2) using `tokenizer.encode` directly instead of `tokenizer.tokenize_messages` in TextCompletion

I'm cool with SFTDataset tho

Comment (joecummings) [reply torchtune/datasets/_finetune.py]: Let's get some other opinions. 

@felipemello1 @SalmanMohammadi @pbontrager @ebsmothers

Comment (SalmanMohammadi) [reply torchtune/datasets/_finetune.py]: m-m-m-m-milord? you asked for me?

SupervisedDataset? TaskDataset? SupervisedTaskDataset? STDataset? 

otherwise +1 for something slightly more descriptive than `FinetuneDataset` - not to be obtuse but, what else are you going to do with it?

File: torchtune/datasets/_finetune.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    1. Dataset-specific transform. This is typically unique to each dataset and extracts
+       the necessary columns into torchtune's :class:`~torchtune.data.Message` format,
+       a standardized API for all model tokenizers.
+    2. If specified, apply a prompt template for the task you are fine-tuning for.
+    3. Model-specific transform or tokenization
+
+    All datasets are formatted into :class:`~torchtune.data.Message`s because for
+    fine-tuning, datasets can be considered as "conversations" with the model,
+    or AI assistant. Thus, we can standardize all text content as messages in a conversation assigned to
+    a :class:`~torchtune.data.Role`:
```
Comment (joecummings) [inline torchtune/datasets/_finetune.py:34]: This isn't rendering correctly.

Comment (RdoubleA) [reply torchtune/datasets/_finetune.py:34]: oh no

File: torchtune/datasets/_finetune.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    For any custom dataset, use the ``message_transform`` to contain all pre-processing to
+    return the list of messages.
+
+    Any model-specific pre-processing that needs to happen can be configured with the ``model_transform``
+    parameter. This is another callable class that contains any custom logic tied to the
+    model you are fine-tuning and will carry over to inference. For example, text + image
+    multimodal datasets requires processing the images in a way specific to the vision
+    encoder being used by the model and is agnostic to the specific dataset.
+
+    Tokenization is handled by the ``model_transform``. All :class:`~torchtune.modules.tokenizers.ModelTokenizer`s
```
Comment (joecummings) [inline torchtune/datasets/_finetune.py:70]: :class:`~torchtune.modules.tokenizers.ModelTokenizer` is not rendering correctly :(

File: torchtune/datasets/_finetune.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    can be treated as a ``model_transform`` since it uses the model-specific tokenizer to
+    transform the list of messages outputted from the ``message_transform`` into tokens
+    used by the model for training. Text-only datasets will simply pass the :class:`~torchtune.modules.tokenizers.ModelTokenizer`
```
Comment (joecummings) [inline torchtune/datasets/_finetune.py:73]: Not rendered.

File: torchtune/datasets/_finetune.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        model_transform (Transform): callable that applies model-specific pre-processing to the sample after the list of
+            messages is created from ``message_transform``. This includes tokenization and any modality-specific
+            transforms. It is expected to return at minimum ``"tokens"`` and ``"mask"`` keys.
+        prompt_template (Optional[PromptTemplate]): template used to format the messages based on their role. This is used
+            to add structured text around the actual messages. The structured text is used in three scenarios:
+
+            - Task-specific templates to gear models for a particular task that it will expect after training
+            - Model-specific templates that are required whenever the model is prompted, such as the [INST]
+              tags in Llama2 and in Mistral
+            - Community standardized templates, such as :class:`~torchtune.data.ChatMLTemplate`
```
Comment (joecummings) [inline torchtune/datasets/_finetune.py:94]: Not rendered

File: torchtune/datasets/_finetune.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    Tokenization is handled by the ``model_transform``. All :class:`~torchtune.modules.tokenizers.ModelTokenizer`s
+    can be treated as a ``model_transform`` since it uses the model-specific tokenizer to
+    transform the list of messages outputted from the ``message_transform`` into tokens
+    used by the model for training. Text-only datasets will simply pass the :class:`~torchtune.modules.tokenizers.ModelTokenizer`
+    into ``model_transform``.
+
+    Args:
+        source (str): path to dataset repository on Hugging Face. For local datasets,
+            define source as the data file type (e.g. "json", "csv", "text") and pass
+            in the filepath in ``data_files``. See Hugging Face's ``load_dataset``
```
Comment (joecummings) [inline torchtune/datasets/_finetune.py:79]: Can we make this a hyperlink?

File: torchtune/datasets/_finetune.py [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+            - Task-specific templates to gear models for a particular task that it will expect after training
+            - Model-specific templates that are required whenever the model is prompted, such as the [INST]
+              tags in Llama2 and in Mistral
+            - Community standardized templates, such as :class:`~torchtune.data.ChatMLTemplate`
+
+            The extra text will still get tokenized as normal text, not as special tokens.
+        filter_fn (Optional[Callable]): callable used to filter the dataset prior to any pre-processing. See
+            the Hugging Face `docs <https://huggingface.co/docs/datasets/v2.20.0/process#select-and-filter>`_ for more
+            details.
+        **load_dataset_kwargs (Dict[str, Any]): additional keyword arguments to pass to ``load_dataset``.
```
Comment (joecummings) [inline torchtune/datasets/_finetune.py:100]: Hyperlink

File: torchtune/datasets/_finetune.py (hunk 10 above)
Comment (RdoubleA) [reply torchtune/datasets/_finetune.py:94]: this is weird, it was just rendering correctly before I made changes. let me go through the built docs again and update

File: torchtune/datasets/_finetune.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    """
+
+    def __init__(
+        self,
+        *,
+        source: str,
+        message_transform: Transform,
```
Comment (joecummings) [inline torchtune/datasets/_finetune.py:107]: For some reason, Transform is not being picked up. Is it included in the docs somewhere?

I'm guessing no...

**Review by joecummings — COMMENTED** [review]
I am the King of Nits.

File: torchtune/datasets/_finetune.py (hunk 8 above)
Comment (SalmanMohammadi) [reply torchtune/datasets/_finetune.py:70]: because we don't actually generate the doc for that anywhere perhaps?

File: tests/torchtune/datasets/test_finetune_dataset.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+        ]
+        prompt_lengths = (14, 4)
+        expected_labels = [
+            [CROSS_ENTROPY_IGNORE_IDX] * prompt_lengths[0]
+            + [10, 3, 7, 2, 4, 2, 3, -1]
+            + [CROSS_ENTROPY_IGNORE_IDX] * prompt_lengths[1]
+            + [10, 1, 6, -1]
+        ]
+        ds = SFTDataset(
+            source="iam/agoofy/goober",
```
Comment (SalmanMohammadi) [inline tests/torchtune/datasets/test_finetune_dataset.py:101]: nice

**Review by joecummings — APPROVED** [review]
This looks world-class - docs are really clear and easy to understand. I'm excited for this to land 💯

File: torchtune/data/_instruct_templates.py [hunk 15]
```diff
@@ -130,113 +130,6 @@ def format(
         return prompt
 
 
-class GrammarErrorCorrectionTemplate(InstructTemplate):
```
Comment (ebsmothers) [inline torchtune/data/_instruct_templates.py:133]: 😍

File: torchtune/datasets/__init__.py [hunk 16]
```diff
@@ -8,6 +8,7 @@
 from torchtune.datasets._chat import chat_dataset, ChatDataset
 from torchtune.datasets._cnn_dailymail import cnn_dailymail_articles_dataset
 from torchtune.datasets._concat import ConcatDataset
+from torchtune.datasets._finetune import SFTDataset
```
Comment (ebsmothers) [inline torchtune/datasets/__init__.py:11]: Sorry to be that guy but imo it's a bit confusing that our canonical dataset does not match the pattern of all our other datasets (class name is the capitalized version of the filename). I'm good with the name `SFTDataset`, but then maybe we should rename the file too?

File: torchtune/models/qwen2/_tokenizer.py [hunk 17]
```diff
@@ -363,7 +366,7 @@ def tokenize_messages(
             if messages:
                 last_message_masked = messages[-1].masked
             mask = mask + [last_message_masked]
-        if max_seq_len:
-            tokens = truncate(tokens, max_seq_len, self.eos_id)
-            mask = truncate(mask, max_seq_len, True)
+        if self.max_seq_len:
+            tokens = truncate(tokens, self.max_seq_len, self.eos_id)
+            mask = truncate(mask, self.max_seq_len, True)
```
Comment (ebsmothers) [inline torchtune/models/qwen2/_tokenizer.py:371]: This one doesn't need to add `__call__`/ `Transform`?

File: torchtune/datasets/_samsum.py [hunk 18]
```diff
@@ -32,15 +40,24 @@ def samsum_dataset(
     - If ``train_on_input`` is False, the prompt is masked out (tokens replaced with -100)
 
     Args:
-        tokenizer (ModelTokenizer): Tokenizer used by the model that implements the ``tokenize_messages`` method.
-        source (str): path string of dataset, anything supported by Hugging Face's `load_dataset`.
+        model_transform (Transform): model specific transform to convert a list of messages
+            output by the dataset to tokens. This will always be a :class:`~torchtune.modules.tokenizers.ModelTokenizer`.
```
Comment (ebsmothers) [inline torchtune/datasets/_samsum.py:44]: > This will always be a :class:`~torchtune.modules.tokenizers.ModelTokenizer`

Just curious, why do we not type it as such then?

File: torchtune/data/_messages.py [hunk 19]
```diff
@@ -99,10 +103,52 @@ def text_content(self) -> str:
...
+    (or equivalent fields specified in column_map) to user and assistant messages,
+    respectively. This is useful for datasets that have two columns, one containing
+    the user prompt and the other containing the model response.
+
+    Args:
+        train_on_input (bool): Whether the model is trained on the user prompt or not.
+            Default is False.
+        column_map (Optional[Dict[str, str]]): a mapping to change the expected "input"
+            and "output" column names to the actual column names in the dataset. Default is None,
+            keeping the default "input" and "output" column names.
```
Comment (ebsmothers) [inline torchtune/data/_messages.py:127]: Should give an example in the docstring

**Review by ebsmothers — COMMENTED** [review]
Please don't hate me for only getting around to reviewing this after you landed it. Anyways it looks great

File: torchtune/datasets/__init__.py (hunk 16 above)
Comment (RdoubleA) [reply torchtune/datasets/__init__.py:11]: yeah I missed that. will include that change in the subsequent PRs

File: torchtune/models/qwen2/_tokenizer.py (hunk 17 above)
Comment (RdoubleA) [reply torchtune/models/qwen2/_tokenizer.py:371]: oh good call out
